### PR TITLE
ARM64: Use stlur for volatile stores

### PR DESCRIPTION
**Pull request by EgorBo**
```cs
volatile int x;

void Test() => x = 42;
```
Codegen diff on arm64 v8.4+ (e.g. Apple M2):
```diff
; Assembly listing for method Program:Test()
            stp     fp, lr, [sp, #-0x10]!
            mov     fp, sp
            mov     w1, #42
-           dmb     ish
-           str     w1, [x0, #0x08]
+           stlur   w1, [x0, #0x08]
            ldp     fp, lr, [sp], #0x10
            ret     lr
```

### CODE REVIEW

**Comment by EgorBo** [comment]
@TIHan @dotnet/jit-contrib PTAL, it's a follow-up to https://github.com/dotnet/runtime/pull/89681 but for stores.

TLDR: on arm64-v8.4 we can use stlur instead of strl or str+dmb because the former now supports IMM offsets. stlur, just like ldapur, has acquire-release semantics which matches our memory model expectations.

**Comment by EgorBo** [comment]
also, cc @VSadov

File: src/coreclr/jit/codegenarm64.cpp [hunk 1]
```diff
@@ -4209,26 +4209,55 @@ void CodeGen::genCodeForStoreInd(GenTreeStoreInd* tree)
         {
             bool addrIsInReg   = addr->isUsedFromReg();
             bool addrIsAligned = ((tree->gtFlags & GTF_IND_UNALIGNED) == 0);
+            bool handledByRcpc = false;
 
-            if ((ins == INS_strb) && addrIsInReg)
+            // On arm64-v8.4+ we can use stlur* instructions with acquire/release semantics
+            // if the address is LEA with just imm offset (unscaled)
+            if (compiler->compOpportunisticallyDependsOn(InstructionSet_Rcpc2) && addrIsAligned &&
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenarm64.cpp:4216]: Maybe you should make this symmetric with the ldapur implementation, and only require `addrIsAligned` for the non-byte cases (INS_strh, INS_str)?

File: src/coreclr/jit/codegenarm64.cpp [hunk 2]
```diff
@@ -4209,26 +4209,55 @@ void CodeGen::genCodeForStoreInd(GenTreeStoreInd* tree)
...
+                (tree->Addr()->OperIs(GT_LEA) && !tree->HasIndex() && (tree->Scale() == 1) &&
+                 emitter::emitIns_valid_imm_for_unscaled_ldst_offset(tree->Offset())))
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenarm64.cpp:4218]: Looks like an unnecessary extra level of parens here.
```suggestion
                tree->Addr()->OperIs(GT_LEA) && !tree->HasIndex() && (tree->Scale() == 1) &&
                emitter::emitIns_valid_imm_for_unscaled_ldst_offset(tree->Offset()))
```

File: src/coreclr/jit/codegenarm64.cpp (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/jit/codegenarm64.cpp]: thanks, unified code with loads to have a single method to decide which load/store instruction to choose.

**Comment by jakobbotsch** [comment]
Will this need a fix similar to #91668?

**Comment by EgorBo** [comment]
> Will this need a fix similar to #91668?

I'd guess - yes? since it's the same ISA. 

```
//  stlur  1X011001000iiiii iiii00nnnnnttttt  9900 0000   [Xn imm(-256..+255)] ARMv8.4 RCPC2
//  stlurb 00011001000iiiii iiii00nnnnnttttt  1900 0000   [Xn imm(-256..+255)] ARMv8.4 RCPC2
//  stlurh 01011001000iiiii iiii00nnnnnttttt  5900 0000   [Xn imm(-256..+255)] ARMv8.4 RCPC2
```
doc: https://developer.arm.com/documentation/ddi0602/2023-06/Base-Instructions/STLUR--Store-Release-Register--unscaled--?lang=en

can you add them as well?

**Comment by jakobbotsch** [comment]
> doc: https://developer.arm.com/documentation/ddi0602/2023-06/Base-Instructions/STLUR--Store-Release-Register--unscaled--?lang=en
> 
> can you add them as well?

Done (but untested)

**Comment by EgorBo** [comment]
@BruceForstall PTAL refactored changes

File: src/coreclr/jit/codegenarmarch.cpp [hunk 3]
```diff
@@ -1728,6 +1728,122 @@ void CodeGen::genCodeForIndexAddr(GenTreeIndexAddr* node)
...
+//
+// Arguments:
+//    currentIns   - the current instruction to perform load/store
+//    indir        - the indirection node representing the volatile load/store
+//    needsBarrier - OUT parameter. Set to true if an explicit memory barrier is required.
+//
+// Return Value:
+//    instruction to perform the volatile load/store with.
+//
+instruction CodeGen::genGetVolatileLdStIns(instruction currentIns, GenTreeIndir* indir, bool* needsBarrier)
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenarmarch.cpp:1743]: Couldn't this function also do the `genIsValidIntReg(targetReg)` check instead of the callers?

**Review by BruceForstall — APPROVED** [review]
LGTM. One question/suggestion.

**Pull request merged**